pkg/controller/controller.go

```diff
@@ -151,45 +173,84 @@ func (c *Controller) handlePodObject(obj interface{}, verb string) {
+		return errors.New("Timeout waiting for pod cache sync")
+	}
+	if v.Type() == reflect.TypeOf(uploadController) {
+		if ok := cache.WaitForCacheSync(stopCh, controller.(*UploadController).servicesSynced); !ok {
+			return errors.New("Timeout waiting for service cache sync")
+		}
+	}
+	glog.V(3).Infof("%s cache has synced", controllerName)
+	return nil
+}
+
 //Run is being called from cdi controllers
-func (c *Controller) run(threadiness int, stopCh <-chan struct{}, controller interface{}) error { //*CloneContorler
+func (c *Controller) run(threadiness int, stopCh <-chan struct{}, controller interface{}) error {
 	defer func() {
+		runtime.HandleCrash()
 		c.queue.ShutDown()
 	}()
-	glog.V(3).Infoln("Starting cdi controller Run loop")
+
 	if threadiness < 1 {
 		return errors.Errorf("expected >0 threads, got %d", threadiness)
 	}
 
-	if !cache.WaitForCacheSync(stopCh, c.pvcInformer.HasSynced) {
-		return errors.New("Timeout waiting for pvc cache sync")
-	}
-	if !cache.WaitForCacheSync(stopCh, c.podInformer.HasSynced) {
-		return errors.New("Timeout waiting for pod cache sync")
+	v := reflect.ValueOf(controller)
+	var importController *ImportController
+	var cloneController *CloneController
+	var uploadController *UploadController
+	if v.IsValid() {
+		if v.Type() == reflect.TypeOf(importController) {
+			waitForCacheSync("ImportController", stopCh, c, v, controller, nil)
+		} else if v.Type() == reflect.TypeOf(cloneController) {
+			waitForCacheSync("CloneController", stopCh, c, v, controller, nil)
+		} else if v.Type() == reflect.TypeOf(uploadController) {
+			glog.V(2).Infoln("Getting/creating certs")
+			ownerController := controller.(*UploadController)
+			if err := ownerController.initCerts(); err != nil {
+				runtime.HandleError(err)
+				return errors.Wrap(err, "Error initializing certificates")
+			}
+			waitForCacheSync("UploadController", stopCh, c, v, controller, uploadController)
+		}
+	} else {
+		return errors.Errorf("controller type is not valid: %v", v)
 	}
-	glog.V(3).Infoln("Controller cache has synced")
+
 	for i := 0; i < threadiness; i++ {
-		//Go is not pure object oriented language. The command repetition below is a result of that.
-		switch t := controller.(type) {
-		case *ImportController:
-			go wait.Until(t.runPVCWorkers, time.Second, stopCh)
-		case *CloneController:
-			go wait.Until(t.runPVCWorkers, time.Second, stopCh)
+		if v.Type() == reflect.TypeOf(importController) {
+			ownerController := controller.(*ImportController)
+			go wait.Until(ownerController.runPVCWorkers, time.Second, stopCh)
+		} else if v.Type() == reflect.TypeOf(cloneController) {
+			ownerController := controller.(*CloneController)
+			go wait.Until(ownerController.runPVCWorkers, time.Second, stopCh)
+		} else if v.Type() == reflect.TypeOf(uploadController) {
+			ownerController := controller.(*UploadController)
+			go wait.Until(ownerController.runWorker, time.Second, stopCh) //<------ change runWorker to runPVCWorkers
```

Comment:
If nothing else I like that this makes the code a bit more explicit so you don't need your comment explaining why the seemingly same calls are repeated like you had to do in the old code.  I'm curious though, is there a reason you switched from using a switch to the if/else-if format here?

Reply:
Rather than doing this type switching in a 'base' class, I think it would be much cleaner for the derived or 'super' controllers to implement a set of interfaces that the base can work with.  For example, the above code would turn into:

```
type WorkerRunner {
    RunWorkers()
}

// super controllers implement RunWorkers() (can prob jest rename existing functions)

ownerController, ok := controller.(WorkerRunner)
// maybe the super controller doesn't want to implement this method for some reason
if ok {
    go wait.Until(ownerController.RunWorkers, time.Second, stopCh)
}

```